v2/client/client.go

```diff
@@ -79,28 +53,54 @@ func (c *Client) SetBasicAuth(username, password string) {
 	c.password = password
 }
 
-// NewClient creates a new Client set to connect to the passed
-// hostname.
-func NewClient(hostname string) (*Client, error) {
-	var err error
+func (c *Client) URL() *url.URL {
+	// Copy our URL, so we are not exposing the original.
+	urlCopy := (*c.url)
+	return &urlCopy
+}
 
-	if "" == hostname {
-		return nil, errNoHost
+func (c *Client) newVoucherRequest(ctx context.Context, url string, image reference.Canonical) (*http.Request, error) {
+	voucherReq := voucher.Request{
+		ImageURL: image.String(),
 	}
 
-	hostnameURL, err := url.Parse(hostname)
-	if nil != err {
-		return nil, fmt.Errorf("could not parse voucher hostname: %s", err)
+	var buf bytes.Buffer
+	if err := json.NewEncoder(&buf).Encode(voucherReq); err != nil {
+		return nil, err
 	}
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, url, &buf)
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Set("Content-Type", "application/json")
+	if c.username != "" && c.password != "" {
+		req.SetBasicAuth(c.username, c.password)
+	}
+	return req, nil
+}
 
-	if "" == hostnameURL.Scheme {
-		hostnameURL.Scheme = "https"
+func (c *Client) doVoucherRequest(ctx context.Context, url string, image reference.Canonical) (*voucher.Response, error) {
+	req, err := c.newVoucherRequest(ctx, url, image)
+	if err != nil {
+		return nil, fmt.Errorf("could create voucher request: %w", err)
+	}
+	resp, err := c.httpClient.Do(req)
+	if nil != err {
+		return nil, err
 	}
+	defer resp.Body.Close()
```

Comment:
cc https://github.com/grafeas/voucher/issues/20 , but not closing that issue.
I think this would manifest as a leak in voucher's clients (e.g. `voucher_subscriber`, not necessarily `voucher_server`).